### DISCUSSION TITLE
Inpainting

### DISCUSSION THREAD

**Discussion by RandomKaos999** [painebenjamin/app.enfugue.ai#118] [Q&A]
For the life of me, I can't seem to figure out inpainting - and based on what it says on the readme, it should be around here... somewhere.

Have you got any guides/videos/documentation for it?

**Comment by painebenjamin** [comment]
Hello! I'm sorry you're having trouble. I'm in the process of creating much more comprehensive documentation as we speak, it'll hopefully be available sometime next week.

For now, this is how Inpainting works in 0.3.1. Unfortunately I'm on in an airport on my way to vacation and not near an instance of enfugue v0.3.1 at the moment so I don't have up-to-date screenshots, but I'll get some from 0.3.0 and explain the difference.

1. Drag an drop an image onto the canvas _or_ paste an image from your clipboard while the enfugue tab is focused _or_ click "Add Image Layer" from the bottom right-hand corner.
2. Under "Visibility Options," make the image _visible_. There is a second option here, _denoised_, do _not_ select this as it will treat your whole image as if it should be inpainted.
3. The inpainting menu now appears on the left-hand-side (this will appear if there is any visible/denoised image on the canvas.) Outpainting is enabled by default, which means that any blank areas on the canvas will be filled in. Enable "Inpainting" in addition to or instead of outpainting allows you to specifically paint in where you want filled in. The inpainting menu bar appears below the model selector, with various brush options. Hover over these to get a tooltip describing what they do.

Hopefully this screenshot helps:

<img width="1295" alt="Screenshot 2023-11-27 at 1 49 23 PM" src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/01d25a42-39fc-4347-80d0-d06136f7da72">

If it helps to understand how inpainting/outpainting/the canvas work in general, this is the logic of how I construct a mask:
1. First, assume no part of the image should be filled (completely black.)
3. If outpainting is enabled and there are empty spaces, fill those empty spaces with white in the mask.
4. If any images exist that are fully denoised, fill the images alpha channel with white in the mask.
5. If the user enabled inpainting, overlay the entire users inpainted area over the mask.

I then evaluate the mask; if there are no filled regions (i.e. it is still completely black) I raise an error, if it is completely denoised (i.e. the mask is completely white,) inpainting is effectively disabled and it gets run through a regular img2img pipeline. Here's a screenshot of the error when I didn't inpaint anything:

<img width="1293" alt="Screenshot 2023-11-27 at 1 43 57 PM" src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/95b5ccaf-7740-4276-b436-abf66d1fec6b">

Once you get your result, click "Edit Image" if you want to inpaint that result, leaving settings as they are will let you just go ahead and paint a new mask as you please.
<img width="1299" alt="Screenshot 2023-11-27 at 1 53 39 PM" src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/31cd6e2b-154d-4a46-96c3-d6ca4fb890aa">

**Reply by RandomKaos999** [reply]
Ah, so i'm not totally nuts. That's a start.

Is there no option for inpainting an image that is generated on the canvas, I would have to save the image first, then drag it on to inpaint it? is it only for images that you drag and drop, or add via the layer button on the lower right?

**Reply by painebenjamin** [reply]
Check the very last section of that comment - click "Edit Image" on any generated image to send it back to the canvas. The menu for working with samples appears in the top-right-hand corner of the sample, including color adjustments, copying to clipboard, and some filters. You can also do a quick upscale there, which is just a shorthand for re-invoking the engine with upscaling enabled and the generated image back on the canvas.

(For what it's worth, that menu in the top-righty-hand corner of a sample used to also appear in the header bar, but evidently that got broken at some point. I'll fix that for v0.3.2.)

**Comment by RandomKaos999** [comment]
Gotcha, spotted and ill have a tinker.

I did attempt to find it, and did try clicking on every option on the canvas including the Edit Image. It didnt occur to me to look on the left side. I was looking on the right as that was the only side that seemed to visibly change when I pressed it.

However, found, all good. Cheers.
